### ISSUE TITLE
Multiple queries in one request - _msearch/

### ISSUE THREAD

**Issue by zauzaj** [elastic/elasticsearch-rails#709]
Hi,
I'm wondering if there is an option to use `/_msearch` using a gem and how? I need this in order to parse results to diff sections and would like to do it via ES rather then parse results in Ruby or to send few diff requests. 

Thanks

**Comment by zauzaj** [comment]
Is there anyone who knows something about this ? @karmi maybe you ?

**Comment by karmi** [comment]
@zauzaj, sorry, I had to focus on another project for the last couple of weeks. 

In general, any ES API can be used directly, since the client can be accessed by `MyModel.__elasticsearch__.client`. But your use case might require something else, can you be more specific in what you're trying to achieve, ideally with a code example?

**Comment by zauzaj** [comment]
Hey @karmi thank you for response. My concrete requirements are next: 
4 diff sections with data from elasticsearch, where every section has to have sub-headline. 
`1st is only users in specific region with availabilities`
`2nd is only users around that region with availabilities ~ 20km`
`3th is only users around regions without availabilities ~ 20km`
`4th not confirmed users ( another type of user) around the regions ~ 20km`

We also have pagination on 20 records per page.

So it's actually 4 very simple queries, which could be written even all together but because of very specific sorting and those subheadlines, mostly, it can't be written in one query.

So that's why I supposed to have multiple queries and to handle all of them, if that's possible. 

If you might have some another suggestion I'm open for any :)

Best Regards

**Comment by karmi** [comment]
Right, thanks for the details!

First of all, as I've said in the last reply, you can call the "Msearch" API through the model, by accessing the client. This, however, gives you the "raw" result from the `elasticsearch` gem, and you would presumably have to do some kind of wrapping on the application layer, so you can present the results easily.

Considering the use case, I don't think "Msearch" is the right approach here, though -- it would be more applicable when you would need to query heterogenous data, let's say from different indices, and would like to present them on one page.

There are at least two approaches which come to my mind.

1/ Have you researched the ["Function Score"](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-function-score-query.html) query? It allows you to manipulate the score (and therefore the order of results) by an algorithm, which looks like something applicable in your case (document has certain properties like "availabilities" and a geographical distance). There's a pretty complex example in the `elasticsearch-dsl` gem [README](https://github.com/elastic/elasticsearch-ruby/tree/master/elasticsearch-dsl#features-overview).

2/ Elasticsearch has recently added a ["Field Collapsing"](https://www.elastic.co/guide/en/elasticsearch/reference/5.x/search-request-collapse.html) feature, which allows you to group results on a certain `keyword` or `numeric` field, which might be applicable in your case as well -- although I'm fairly sure you'd have to prepare the data for it during indexing time.

I'm interested in hearing which approach seems to be most practical in your case.

**Comment by zauzaj** [comment]
Hi @karmi,
I solved it on different way, as non of the above suggestions not fully matched our requirements. So what I did are 4 diff strategies for 4 diff sections ( which make query to ES accordingly) and then using ruby have recursive function which call those sections. 
Also I've implemented our custom pagination for those sections, so before calling next section we check if there is already 20 results per page and if true return results. So from technical side it's not so perfect, but from performance point it works really nice. 
Anyway thanks a lot on your suggestions, I learned more new things during investigation as well :)

Best regards
